// node-opcua-nodeset-ua/dist/enum_trust_list_masks.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumTrustListMasks = void 0;
/**
 * |           |                                                            |
 * |-----------|------------------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                                |
 * | nodeClass |DataType                                                    |
 * | name      |TrustListMasks                                              |
 * | isAbstract|false                                                       |
 */
var EnumTrustListMasks;
(function (EnumTrustListMasks) {
    /**
     * No fields are provided.
     */
    EnumTrustListMasks[EnumTrustListMasks["None"] = 0] = "None";
    /**
     * The TrustedCertificates are provided.
     */
    EnumTrustListMasks[EnumTrustListMasks["TrustedCertificates"] = 1] = "TrustedCertificates";
    /**
     * The TrustedCrls are provided.
     */
    EnumTrustListMasks[EnumTrustListMasks["TrustedCrls"] = 2] = "TrustedCrls";
    /**
     * The IssuerCertificates are provided.
     */
    EnumTrustListMasks[EnumTrustListMasks["IssuerCertificates"] = 4] = "IssuerCertificates";
    /**
     * The IssuerCrls are provided.
     */
    EnumTrustListMasks[EnumTrustListMasks["IssuerCrls"] = 8] = "IssuerCrls";
    /**
     * All fields are provided.
     */
    EnumTrustListMasks[EnumTrustListMasks["All"] = 15] = "All";
})(EnumTrustListMasks = exports.EnumTrustListMasks || (exports.EnumTrustListMasks = {}));
